### ISSUE TITLE
Empty values are not getting saved

### ISSUE THREAD

**Issue by vengi83644** [OrchardCMS/OrchardCore#11373]
I know this is weird. But this happens in the Orchard 1.2.2 version.

Repro,
1. Create a content type with field - a simple textbox and set as not required.
2. Create content for this new content type filling some text in the textbox. Publish it.
3. Now, edit the content and remove all the text from the textbox and publish it.
4. If you open the content now, the text is not removed.

Expected is that the textbox should be empty.

The same goes for checkboxes when there is only one checkbox and unchecking it.

Thanks,
R. Venkatesan

**Comment by hishamco** [comment]
Is this happen in the latest bits too? If Yes it's a bug

**Comment by hishamco** [comment]
> Are there any breaking changes other than .net core support? 

AFAIK, there's no breaking in this release except changing the TFM

**Comment by vengi83644** [comment]
This happens in 1.3.0 too.

I tested this in the text field, link field, and checkboxes. 

Repro,
1. Setup Orchard in Decoupled CMS mode.
2. Create a content type with a text field.
3. Create content for this new content type by adding any text in that field and publish it.
4. Now, edit the content and remove the contents in the text field and publish it.
5. Again, go to edit page and you will see the contents that you entered initially and not removed in the Step #4

**Comment by vengi83644** [comment]
I tried debugging and found the problem is in the DisplayDriver of each fields

Here is an example that's easy for me to explain.

Take this BooleanFieldDisplayDriver,

The problem I faced in boolean fields is we were able to set that to true and save it. But when we want to uncheck it and save, it won't change and remain true always.

So, I created a Module and copied the "BooleanFieldDisplayDriver" from OC code and added it to my module to debug it. 

model.Value is getting set with the latest value(in my case its "false") and the model.Field is still having true and this is why its not updating in the database. 

![image](https://user-images.githubusercontent.com/5276778/169625977-0fb87e84-91a2-4bbb-bd51-1ca204dad757.png)

I tried setting the model.Field.Content.Value to the latest value and its working - I checked in the database.

This is the same case for Link field, Text field and other fields as well.

Thanks,
R. Venkatesan

**Comment by vengi83644** [comment]
This is what worked for me. I am using version 1.3 and I am going to use this for now.


Added new module and using my own display drivers copied from version 1.3
![image](https://user-images.githubusercontent.com/5276778/169712715-76286ca9-70eb-4a30-bb8b-327988250d9a.png)

Text Field
![image](https://user-images.githubusercontent.com/5276778/169712616-5f076bb1-5672-40a6-ae30-dbaafc8b3f5d.png)

Bool Field
![image](https://user-images.githubusercontent.com/5276778/169712636-5f06a9b1-45e4-4c58-89d2-2a39d4cb49dd.png)

Link Field
![image](https://user-images.githubusercontent.com/5276778/169712661-0788ff2e-d23f-400e-a5f6-232780749c7a.png)

**Comment by Skrypt** [comment]
![image](https://user-images.githubusercontent.com/3228637/170067455-9054d03d-3d8e-4d64-acd5-c7935b8eda6b.png)

Here is the proof.

![image](https://user-images.githubusercontent.com/3228637/170067715-54c869c5-32f9-43a6-841c-918e46597da6.png)

**Comment by vengi83644** [comment]
Hey @Skrypt , Thanks for the quick response. 

I tried with the sample App that is in Orchard Source. I didn't get this issue in the sample. I agree this is not in Orchard source.

But when I import the content definition from my own app to the sample app, I get the issue. I don't do any other changes to the sample app.

Would my own content types and content parts trigger this issue?

Thanks,
R. Venkatesan

**Comment by Skrypt** [comment]
Yes, there may be an issue with your Content Definitions. Try to compare them with a Content Type exported from source code solution. If there are differences then let us know. Normally, it should have been migrated if the issue is that you upgraded from a previous version.

**Comment by vengi83644** [comment]
@Skrypt 
Thanks. I figured out the problem.

I am using the below JsonSerializerSettings as default in my app.

```
JsonConvert.DefaultSettings = () =>
            {
                var settings = new JsonSerializerSettings()
                {
                    NullValueHandling = NullValueHandling.Ignore,
                    DefaultValueHandling = DefaultValueHandling.Ignore
                };

                return settings;
            };
```

This conflicts with the settings(JsonMergeSettings I hope) in OC.

Please let me know your suggestions to overcome this.

Thanks,
R. Venkatesan

**Comment by Skrypt** [comment]
You could try to create a scope that uses these settings only when requested for the parts that you require to use them. Else, changing the defaults of Orchard Core is probably not a good idea.

**Comment by vengi83644** [comment]
OK, but I am using this setting in my MVC app as a default setting for all my JSON serialization. 

I don't want OC to pick this setting as I hope it has its own setting in it. 

Can you please explain how do we scope my setting to my app alone?

**Comment by vengi83644** [comment]
@Skrypt Any inputs?

**Comment by jtkech** [comment]
Hmm, we use a lot Newtonsoft and I don't think we override any System.Text.Json settings at the tenant/modules level, meaning that we use the inherited default settings. So you may need to do revert your custom settings at the tenant level.

We have `Configure()` and `ConfigureServices()` helpers to do it from the app but executed at the tenant/modules level.

    .AddOrchardCms()
    .ConfigureServices(tenantServices =>
    {
        // Some stuff.
    }

**Comment by vengi83644** [comment]
Thanks @jtkech. Let me try and update here.

**Comment by vengi83644** [comment]
Thanks @jtkech @Skrypt for helping me in identifying the root cause,

I have fixed the issue. Here is what worked for me,

I am setting this in my .Net core app
```
JsonConvert.DefaultSettings = () =>
            {
                var settings = new JsonSerializerSettings()
                {
                    NullValueHandling = NullValueHandling.Ignore,
                    DefaultValueHandling = DefaultValueHandling.Ignore
                };

                return settings;
            };
```

and I am setting this in the OC tenant settings,
I have set the null and default handling to "Include" to match with JsonMergeSettings in OC

```
builder.Services.AddOrchardCms()
                .AddMvc()
                .AddDatabaseShellsConfiguration()
                .ConfigureServices((services) =>
                {
                    JsonConvert.DefaultSettings = () =>
                    {
                        var settings = new JsonSerializerSettings()
                        {
                            NullValueHandling = NullValueHandling.Include,
                            DefaultValueHandling = DefaultValueHandling.Include
                        };

                        return settings;
                    };

                    services.ConfigureHtmlSanitizer((sanitizer) =>
                    {
                        sanitizer.AllowedSchemes.Add("mailto");
                    });
                });
```
